Fix #12948 (slow domain join)

damiendoligez opened the pull request:
I think this is what @gadmm had in mind in https://github.com/ocaml/ocaml/issues/12399#issue-1814032987.

fixes #12948 
Also fixes the first item of #12399

damiendoligez committed:
interrupt tick thread without delay on domain shutdown

gadmm commented on runtime/platform.c:
```diff
@@ -114,6 +114,15 @@ void caml_plat_wait(caml_plat_cond* cond)
+  int res;
+  caml_plat_assert_locked(cond->mutex);
+  res = pthread_cond_timedwait(&cond->cond, cond->mutex, deadline);
+  if (res != ETIMEDOUT) check_err("timedwait", res);
```
One documentation for `pthread_cond_timedwait` mentions spurious wakeups. In these cases one should probably restart `pthread_cond_timedwait` ~rather than raise an exception~. (edit: callers should take spurious wakeups into account)

gadmm commented on runtime/platform.c:
```diff
@@ -114,6 +114,15 @@ void caml_plat_wait(caml_plat_cond* cond)
+  caml_plat_assert_locked(cond->mutex);
+  res = pthread_cond_timedwait(&cond->cond, cond->mutex, deadline);
+  if (res != ETIMEDOUT) check_err("timedwait", res);
+  return res;
```
```suggestion
  return res != ETIMEDOUT;
```

damiendoligez replied:
At this point, `res` is either 0 (signal or spurious wake-up) or ETIMEDOUT (deadline reached), so this is just adding a negation. I would rather keep this function as close as possible to `pthread_cond_timedwait` and leave the return value as-is.

gadmm commented on runtime/caml/platform.h:
```diff
@@ -114,8 +114,7 @@ typedef struct { pthread_cond_t cond; caml_plat_mutex* mutex; } caml_plat_cond;
 void caml_plat_wait(caml_plat_cond*);
-/* like caml_plat_wait, but if nanoseconds surpasses the second parameter
-   without a signal, then this function returns 1. */
+int caml_plat_timedwait(caml_plat_cond*, const struct timespec *);
```
```suggestion
/* return 0 on timeout */
int caml_plat_timedwait(caml_plat_cond*, const struct timespec *);
```

gadmm commented on otherlibs/systhreads/st_pthreads.h:
```diff
@@ -297,15 +315,27 @@ static int st_event_wait(st_event e)
+    }else{
+      deadline.tv_sec = curtime.tv_sec;
+    }
+    (void) caml_plat_timedwait (&Tick_thread_control.cond, &deadline);
```
Correcting my previous comment, I'd see spurious wakeups handled here:
```suggestion
    while(caml_plat_timedwait (&Tick_thread_control.cond, &deadline))
    {
      if (Tick_thread_control.state == Tick_stop) goto …;
      /* In case of spurious wakeup, keep waiting */
    };
```

gadmm commented on otherlibs/systhreads/st_pthreads.h:
```diff
@@ -297,15 +315,27 @@ static int st_event_wait(st_event e)
+  caml_plat_lock (&Tick_thread_control.mu);
+  while(1){
+    if (Tick_thread_control.state == Tick_stop) break;
+    gettimeofday (&curtime, NULL);
```
The clock for `caml_plat_timedwait` is set inside `caml_plat_cond_init_aux` in a platform-dependent manner. Please use e.g. `clock_gettime` to make sure the deadline is on the same clock.

damiendoligez replied:
Unfortunately, `clock_gettime` is not available on MacOS, so I had to reproduce the platform-dependent part.

gadmm commented on otherlibs/systhreads/st_stubs.c:
```diff
@@ -572,9 +572,12 @@ CAMLprim value caml_thread_initialize(value unit)
-    atomic_store_release(&Tick_thread_stop, 1);
+    caml_plat_lock (&Tick_thread_control.mu);
+    Tick_thread_control.state = Tick_stop;
+    caml_plat_unlock (&Tick_thread_control.mu);
+    caml_plat_signal (&Tick_thread_control.cond);
```
```suggestion
    caml_plat_signal (&Tick_thread_control.cond);
    caml_plat_unlock (&Tick_thread_control.mu);
```
The `DEBUG_LOCK` seems to enforce that the mutex is locked when entering `caml_plat_signal`. But there is a PR currently under review which aims to remove the needless restriction.

However according to some sources online, schedulers are programmed to better handle the case where `pthread_cond_signal` is called with the mutex locked. Not an expert on these details though. Feel free to ignore this change if you have a strong opinion about it. I am curious to know the rationale for the best choice.

damiendoligez replied:
I have no strong opinion and I am ready to believe that schedulers are optimized that way.

gadmm reviewed (changes requested):
Reading more closely this time around, I changed where I think one should deal with spurious wakeups. I also have added a comment about the choice of clock.

damiendoligez commented:
Many thanks for the thorough review. I changed the code to implement your suggestions.

damiendoligez commented:
Running precheck at https://ci.inria.fr/ocaml/job/precheck/972/ .

gasche commented:
This needs a rebase, it would be useful to get an explicit approval from @gadmm, but also the CI reports a lot of failed tests on Windows:

```

    tests/backtrace/backtrace_systhreads.ml
    tests/backtrace/callstack.ml
    tests/c-api/test_c_thread_has_lock_systhread.ml
    tests/lib-channels/input_all.ml
    tests/lib-systhreads/boundscheck.ml
    tests/lib-systhreads/multicore_lifecycle.ml
    tests/lib-systhreads/test_c_thread_register.ml
    tests/lib-threads/backtrace_threads.ml
    tests/lib-threads/bank.ml
    tests/lib-threads/bufchan.ml
    tests/lib-threads/close.ml
    tests/lib-threads/fileio.ml
    tests/lib-threads/mutex_errors.ml
    tests/lib-threads/pr5325.ml
    tests/lib-threads/pr7638.ml
    tests/lib-threads/pr8857.ml
    tests/lib-threads/prodcons.ml
    tests/lib-threads/prodcons2.ml
    tests/lib-threads/sieve.ml
    tests/lib-threads/swapchan.ml
    tests/lib-threads/tls.ml
    tests/lib-threads/torture.ml
    tests/lib-threads/uncaught_exception_handler.ml
    tests/lib-unix/win-socketpair/test.ml
    tests/parallel/fib_threads.ml
    tests/parallel/multicore_systhreads.ml
    tests/parallel/test_c_thread_register.ml
    tests/regression/pr12948/test.ml
    tests/statmemprof/blocking_in_callback.ml
    tests/statmemprof/moved_while_blocking.ml
    tests/statmemprof/thread_exit_in_callback.ml
```

This probably needs some investigation if the results are consistent after the rebase and CI rerun.

gadmm commented:
I am happy to do a final round of review after a rebase & resolution of CI failures.

gasche commented:
@damiendoligez I wonder if you have an idea of what is going on with the Windows stuff.

edwintorok commented on runtime/platform.c:
```diff
@@ -115,6 +139,15 @@ void caml_plat_wait(caml_plat_cond* cond)
+{
+  int res;
+  caml_plat_assert_locked(cond->mutex);
+  res = pthread_cond_timedwait(&cond->cond, cond->mutex, deadline);
```
Are you sure this is correct? The condition var has an `&`, but the mutex doesn't.

edwintorok commented:
> I wonder if you have an idea of what is going on with the Windows stuff.

Most of the CI logs are gone, but this one is still here: https://ci.appveyor.com/project/avsm/ocaml/builds/49722435?fullLog=true

```
> Fatal error: cannot load shared library dllthreads
> Reason: flexdll error: cannot relocate caml_plat_fatal_error RELOC_REL32, target is too far: fffffffceb5a62a1  ffffffffeb5a62a1
```

There is some discussion about that here https://cygwin.com/pipermail/cygwin/2020-March/244095.html, which seems to hint that perhaps there are multiple versions of the library, but obviously you can only have one of them loaded at runtime, or that cygwin's autorebase broke them or that opam overwrite some dll. I don't see why that'd suddenly start failing on this PR, but it does introduce a new symbol, so if there is some preexisting problem where it loads the wrong libraries it might explain why it fails.

edwintorok commented:
Is there perhaps some macro missing from the declaration of `caml_plat_timedwait`, like `CAMLextern`?  (same for the other newly introduced functions/dependencies between platform.c and st_pthreads.h)

https://github.com/ocaml/ocaml/pull/11577 seems to hint that missing `CAMLextern` can lead to relocation errors in flexdll (which is the error seen in the CI for this PR), although the macro itself seems to be undocumented, and it is not clear why some functions in `platform.h` use it (e.g. `caml_plat_mutex_init`), and others do not (e.g. `caml_plat_assert_locked`).
At a guess: anything that needs to be used by a different DLL (such as dllthreads) needs a `CAMLextern` in platform.h and other headers, and anything that is used only internally probably does not. (`CAMLextern` seems to expand to `CAMLDLLIMPORT extern`).
Although this also means that if some code in dllthreads (or elsewhere) starts using a symbol from `platform.h` that it didn't before, then that symbol will also need a `CAMLextern`, so this isn't just a problem with newly introduced symbols, it is a problem with newly introduced symbol uses too. Since platform and threading is quite closely tied together should all non-inline functions in platform.h be marked as `CAMLextern`?

damiendoligez commented:
Rebased, but now I need to implement the Windows side of things because of #13416: we don't use pthreads on Windows any more.

damiendoligez committed:
fixes

damiendoligez committed:
Change API and add a windows version.

Windows' equivalent to `ptrhead_cond_timedwait` takes a delay in
milliseconds rather than a deadline in a `struct timespec`, so I
changed the interface of `caml_plat_timedwait` to use the lowest
denominator.

damiendoligez committed:
remove TABs (check-typo)

damiendoligez commented:
@gadmm, I had to change the API of `caml_plat_timedwait` to accomodate Windows and I simplified the tick thread's loop based on the assumption that spurious wakeups are uncommon.
